src/mlpack/methods/ann/gan_impl.hpp

```diff
@@ -0,0 +1,293 @@
+      weightsDiscriminator, 1, false, false);
+
+  for (size_t i = 0; i < generator.network.size(); ++i)
+  {
+    offset += boost::apply_visitor(WeightSetVisitor(std::move(parameter),
+        offset), generator.network[i]);
+
+    boost::apply_visitor(resetVisitor, generator.network[i]);
+  }
+
+  for (size_t i = 0; i < discriminator.network.size(); ++i)
+  {
+    offset += boost::apply_visitor(WeightSetVisitor(std::move(parameter),
+        offset), discriminator.network[i]);
+
+    boost::apply_visitor(resetVisitor, discriminator.network[i]);
+  }
+
+  reset = true;
+}
+
+template<typename Generator, typename Discriminator, typename IntializerType>
+template<typename OptimizerType>
+void GenerativeAdversarialNetwork<Generator, Discriminator, IntializerType>
+::Train(OptimizerType& Optimizer, size_t iterations, size_t k)
+{
+  if (!reset)
+    Reset();
+  for (size_t i = 0; i < iterations; i++)
+  {
+    // intialise all the variable again
+    noiseData.set_size(generatorInSize, batchSize);
+    fakeData.set_size(predictors.n_rows, batchSize);
+    fakeLables.set_size(1, batchSize);
+    tempTrainData.set_size(predictors.n_rows, 2 * batchSize);
+    tempLabels.set_size(1, 2 * batchSize);
+    fakeData.zeros();
+    fakeLables.zeros();
+    size_t offset = 0;
+
+    // Train the discrminator network k times
+    for (size_t i =0; i < k; i++)
+    {
+      if (k * batchSize > predictors.n_cols)
+      {
+        std::cout << "k value too large" << std::endl;
+        return;
+      }
+      // if last batch excess predictors columns
+      if (offset + batchSize >  predictors.n_cols)
+        offset -= (batchSize - predictors.n_cols);
+
+      // Generate fake data for discrminator to train on.
+      Generate(batchSize, std::move(fakeData), std::move(noiseData));
+      fakeLables.zeros(1, batchSize);
+      // Create training data for discrminator
+      tempTrainData.cols(0, batchSize - 1) = arma::mat(fakeData.memptr(),
+          fakeData.n_rows, batchSize, false, false);
+      tempTrainData.cols(batchSize, tempTrainData.n_cols - 1) = arma::mat(
+          predictors.memptr() + offset, predictors.n_rows, batchSize,
+          false, false);
+      tempLabels.cols(0, batchSize - 1) = arma::mat(
+          fakeLables.memptr() + offset, fakeLables.n_rows, batchSize,
+          false, false);
+      tempLabels.cols(batchSize, tempLabels.n_cols - 1) = arma::mat(
+          responses.memptr() + offset, responses.n_rows, batchSize,
+          false, false);
+      offset += batchSize * k;
+      // Training discriminator
+      this->predictors = std::move(tempTrainData);
+      this->responses = std::move(tempLabels);
+      numFunctions = predictors.n_cols;
+      // set here discrminator evaluate function
+      discriminator.predictors = predictors;
+      discriminator.responses = responses;
+      // do one pass over the dataset
+      Optimizer.MaxIterations() = numFunctions;
+      Optimizer.Optimize(*this, parameter);
```

Comment:
The problem with this strategy is that you reset the update policy for each call of ``Optimize``, which is fine for the vanilla update policy, but maybe it's a better idea to generate the complete training set and call ``Optimize`` once?